docs/geps/gep-01.rst

```diff
@@ -0,0 +1,253 @@
+
+  - For column names in the user-facing API, there is a hard limit of 15 characters
+  - For other column names, there is a soft limit of 15 and a hard limit of 20 characters
+
+* The language should generally be English in all coding efforts and documentation.
+  German should be used for all institutional features for precision.
+
+We explain the background for these choices below.
+
+
+Motivation and Scope
+--------------------
+
+Naming conventions are important in order to build and maintain GETTSIM as a coherent
+library. Since many people with different backgrounds and tastes work on it, it is
+particularly important to clearly document our rules for naming things.
+
+There are three basic building blocks of the code:
+
+1. The input and output data, including any values stored intermediately.
+2. The parameters of the tax transfer system, as detailed in the YAML files.
+3. Python identifiers, that is, variables and functions.
+
+The general rules and considerations apply in the same way to similar concepts, e.g.,
+groups of parameters.
+
+
+General considerations
+----------------------
+
+Even though the working language of GETTSIM is English, all of 1. (column names) and 2.
+(parameters of the taxes and transfers system) should be specified in German. Any
+translation of detailed rules---e.g., ....---is likely to lead to more confusion than
+clarity. Non-German speakers would need to look things up, anyhow.
+
+
+Column names (a.k.a. "variables" in Stata)
+------------------------------------------
+
+We impose a hard limit of 15 characters for all column names that are part of the API,
+i.e., those that are an input to or an output of GETTSIM's main simulation functions.
+This is for the benefit of Stata users, who face a strict limit of 32 characters for
+their column names. Furthermore, where developers using other languages may store
+different experiments in different variables, Stata users' only chance to distinguish
+them is to append characters to the column names.
+
+If a column is only for internal use, it should start with an underscore.
+
+Even though not implemented at the time of this writing, we plan to allow users to pass
+in English column names and get English column names back. Potentially also standardised
+variables like in `EUROMOD <https://www.euromod.ac.uk/>`_ or the ... standard.
+
+Parameters of the taxes and transfers system
+--------------------------------------------
+
+[merge this with @mjbloemer's parts below]
+
+* The parameters are stored by group (e.g., ``arbeitsl_geld``, ``kinderzuschlag``).
+  These groups or their abbreviations should not re-appear in the name.
+* Use Python containers like tuples and namedtuples (the mutable versions of these,
+  i.e., lists and dictionaries, might be better known) where relevant.
+
+E.g., instead of ``kgeld1: 204``, ``kgeld2: 204``, ``kgeld3: 210``, and ``kgeld4: 235`` in the file ``kindergeld.yaml``, use
+
+    .. code-block:: yaml
+
+        beträge:
```

Comment:
Could we avoid umlauts? Difficult to enter on most non-German keyboards.

Reply:
I have a pretty strong opinion here and it is pro-Umlaut. 

One of the founding principles of Python is "Code is read much more often than it is written." This is particularly true here. We would deviate from much better readability (and uniqueness; in English it is never that obvious whether ö will become oe or o) for sake of avoiding a minor annoyance for very few people in very few circumstances -- this code is just not written from scratch all that often.

In all likelihood, the vast majority of GETTSIM developers will have access to a German keyboard. And anybody else can use the replacement and ask for somebody else to change it (as a case in point, see [here](https://github.com/iza-institute-of-labor-economics/gettsim/pull/50#discussion_r364433143).

Reply:
Hmmm... `bezugsgröße_ost`? I see no big advantage in readability. I guess I am used to `ae` replacements.

So we specifically allow umlauts (and sharp s?) but require all lowercase words and a lower bar `_` in place for spaces?

And I think being open for non German speakers and users is a very valid point (sonst können wir das hier alles auch auf Deutsch machen. Dann wären einige Diskussionen um das deutsche Steuer- und Transfersystem weniger umständlich). I think writing code or addressing/discussing functions parameter names in mails or github issues for instance is part of that. 

When browsing through OpenFisca I have to look up French words pretty often (they don't use accents btw, `._impot`, `._bareme`,...). I'm not sure what is the right way but it could be good to use English language for clear cases such as vat or income tax.

Plus aren't there some issues with Umlaut Handling in other programming languages? Let's say someone wants to analyze the param database in a different language or re-use/share the param database with the Stata model?

Reply:
> Hmmm... `bezugsgröße_ost`? I see no big advantage in readability. I guess I am used to `ae` replacements.

Yes, I find "bezugsgröße" much more readable than "bezugsgroesse". The "ost" part will be solved differently, anyhow.

> So we specifically allow umlauts (and sharp s?) but require all lowercase words and a lower bar `_` in place for spaces?

That depends on whether we are sure we never need these guys as Python identifiers. E.g. `arbeitsl_geld_2.beträge` vs. `arbeitsl_geld["Beträge"]`. As long as we want to keep the former an option (which I think we should), Python conventions it is. 

> And I think being open for non German speakers and users is a very valid point (sonst können wir das hier alles auch auf Deutsch machen. Dann wären einige Diskussionen um das deutsche Steuer- und Transfersystem weniger umständlich). I think writing code or addressing/discussing functions parameter names in mails or github issues for instance is part of that.

Who would mind someone being a bit sloppy in github issues /  Zulip discussions? 

> When browsing through OpenFisca I have to look up French words pretty often (they don't use accents btw, `._impot`, `._bareme`,...). I'm not sure what is the right way but it could be good to use English language for clear cases such as vat or income tax.

The bigger problem is that English translations refer to concept rather than identifiers for laws. E.g., the Solidaritätssteuer would have to be part of income tax even though it is not part of Einkommensteuer. How to distinguish between Elterngeld, Elterngeld +, and Erziehungsgeld? Etc. I would choose one language to avoid discussing each time we add a new parameter whether it is in German or English. But this is somewhat orthogonal to the ASCII / non-ASCII discussion.

> Plus aren't there some issues with Umlaut Handling in other programming languages? Let's say someone wants to analyze the param database in a different language or re-use/share the param database with the Stata model?

There are good converters. This is a very distant possibility; I really would not want to make design decisions based on it.

Reply:
I'm not convinced by the pro-Umlaut argument. Readability is the only argument if I understand correctly.

I don't think we need to improve readability. Again: And if so, why only Umlauts and sharp-s? I think it would be ridiculous to make use of anything that slightly improves readability. E.g. you could also improve readability of the variable name by using [UTF-8 whitespace characters](http://jkorpela.fi/chars/spaces.html) instead of `_` (assuming not all are reserved by python) or use full stop characters when abbreviating names.

What do you say on lowercase vs. uppercase ([I think you prefer lowercase](https://github.com/iza-institute-of-labor-economics/gettsim/pull/60#pullrequestreview-313397771)) in the light of this readability argument? I have the impression that the readability argument would definitely require uppercase German nouns (please not).



> Code is read much more often than it is written.

That refers to writing clean source code in general, and not specifically to the naming of variables.

Also I think that underestimates how often variable names will be *typed*:
* in github issues, mails and chat, as mentioned before;
* when browsing source code, searching and grepping occurrences of a variable name in and across files

There are also people in Germany that work with US keyboards (some say US keyboards are more comfy for coding).

I think the disadvantages still overweight.
   

> There are good converters.

OK, good to know. Could you point me to one for the mentioned tasks?


> This is a very distant possibility; I really would not want to make design decisions based on it.

I already try to keep the param names of our Stata based model in sync with the new gettsim names to improve maintainability.

Reply:
> I'm not convinced by the pro-Umlaut argument. Readability is the only argument if I understand correctly.

And uniqueness, to some extent. We could of course agree always to use the output of something like https://pypi.org/project/Unidecode/ (re your question on converters) throughout, but that would be more rules to keep in mind... 

> I don't think we need to improve readability. Again: And if so, why only Umlauts and sharp-s? I think it would be ridiculous to make use of anything that slightly improves readability. 

Can we please try to tone things down a little? I think there are two very valid opinions around here and I do not find anything that would be ridiculous. 

> E.g. you could also improve readability of the variable name by using [UTF-8 whitespace characters](http://jkorpela.fi/chars/spaces.html) instead of `_` (assuming not all are reserved by python) or use full stop characters when abbreviating names.

> What do you say on lowercase vs. uppercase ([I think you prefer lowercase](https://github.com/iza-institute-of-labor-economics/gettsim/pull/60#pullrequestreview-313397771)) in the light of this readability argument? I have the impression that the readability argument would definitely require uppercase German nouns (please not).

Well, there are clear rules for Python variable/function names, which are `lowercase_with_underscores`. The moment you see an uppercase identifier, you think it is a class.

Whether we are talking about identifiers or strings in general, it all depends on what is "normal". Using a character that looks like another but is actually different would indeed be a no-go.

There are trade-offs here, I am not saying that there is a completely obvious solution one way or another.

> > Code is read much more often than it is written.
> 
> That refers to writing clean source code in general, and not specifically to the naming of variables.

Yes, of course. And naming is one of them.

> Also I think that underestimates how often variable names will be _typed_:
> 
>     * in github issues, mails and chat, as mentioned before;
> 
>     * when browsing source code, searching and grepping occurrences of a variable name in and across files

The latter is a fair enough point. Usually you will be able to copy & paste, though. 

> There are also people in Germany that work with US keyboards (some say US keyboards are more comfy for coding).

I do. But switching them is a keystroke which I am more than happy to do for readability. 

> > This is a very distant possibility; I really would not want to make design decisions based on it.
> 
> I already try to keep the param names of our Stata based model in sync with the new gettsim names to improve maintainability.

So where are the limits of UTF-8 support in Stata? Assuming that supporting <= 13 is not an issue?

Reply:
> So where are the limits of UTF-8 support in Stata? Assuming that supporting <= 13 is not an issue

Stata seems to use/allow unicode UTF-8 encoding since version 14: https://www.stata.com/stata14/unicode/. you can `generate einkommen_tsd€` or even `gen zufriedenheit_😀️ = 1`.

No other problems with non-ascii-characters? Filesystems? URLs? path specifications?

Reply:
> > So where are the limits of UTF-8 support in Stata? Assuming that supporting <= 13 is not an issue
> 
> Stata seems to use/allow unicode UTF-8 encoding since version 14: https://www.stata.com/stata14/unicode/. you can `generate einkommen_tsd€` or even `gen zufriedenheit_😀️ = 1`.

They did make it to the 21st century, after all :-)

> No other problems with non-ascii-characters? Filesystems? URLs? path specifications?

I am sure there are, but I do not think this is relevant for our naming conventions here? A URL is a URL, we should use the correct one regardless of what is in there. Paths should work in general, but I am not sure I would want to probe that.

Reply:
As long as people are saving their stuff in `utf-8`, this should not be major problem. Might be worth mentioning in the contributor guidelines, even though it should be self-evident nowadays.

Reply:
For me `bezugsgroesse` isn't much less readable than `bezugsgroesse`. 
I am therefore in favour of replacing all German-specific characters.

Reply:
Now that @mjbloemer has started using Unicode all over the place in his Stata code? ;-)

Reply:
Did he? If this is already decided, that's also fine for me!

Reply:
That is how I interpreted his last message :-) But I was mostly joking, shouldn't do that here. Sorry.

Nothing is decided yet. My interpretation is, however, that the mood has become more favourable towards special characters. 

In any case, this will not affect a huge amount of code and it is not the question deciding the future of GETTSIM. I would of course go with the majority.

Somewhat related: Do we all agree that names / descriptions / notes should keep correct spelling? These may end up directly in reports etc., I would not want to rename them before publishing something in German.

Reply:
> Did he?

no of course not. Why would you think so?

Reply:
> > Did he?
> 
> no of course not. Why would you think so?

https://github.com/iza-institute-of-labor-economics/gettsim/pull/60#discussion_r369011217

Yes, I am sorry. I should not joke around.

Reply:
> Somewhat related: Do we all agree that names / descriptions / notes should keep correct spelling? These may end up directly in reports etc., I would not want to rename them before publishing something in German.

Is this still about the special characters or what should be the reason to 
misspell?

Reply:
Yes, but in another context than what we were discussing before — i.e., fields that will never end up in code. 

Checking the stuff I would consider less controversial first 😉

Reply:
Ok, then I am not sure what 

> names / descriptions / notes

refers to.

Reply:
This is all about the contents of the yaml files. See [below](https://github.com/iza-institute-of-labor-economics/gettsim/pull/60#discussion_r369256241), for example.

Our original discussion was about the identifiers of parameters (e.g., `e_anr_frei`) and it is unresolved.

I am now probing the waters first if people have an aversion against special characters per se -- also in settings that have nothing to do with code --- or whether this is more about a reluctance to see them as Python identifiers.

E.g., if I was building a table with parameters, my identifying column would typically be what we include in "name". I would not want to (pay people to) manually fix all occurrences of "Bezugsgroesse" to "Bezugsgröße" etc. and I hope we agree that in published reports, ASCII-normalised spelling is not an option.

Reply:
Ok, thanks for the clarification! 

> Do we all agree that names / descriptions / notes should keep correct spelling?
I agree!